arrows/vxl/high_pass_filter.cxx

```diff
@@ -0,0 +1,490 @@
+        {
+          sum += *pixelS2;
+        }
+
+        // starting boundary case: the kernel width is expanding
+        for(; i < kernel_width; ++i, pixelS2 += istepS, pixelD += istepD )
+        {
+          *pixelD = static_cast< PixType >( sum / i );
+          sum += *pixelS2;
+        }
+
+        // general case: add the leading edge and remove the trailing edge.
+        for(; i < ni;
+            ++i, pixelS1 += istepS, pixelS2 += istepS, pixelD += istepD )
+        {
+          *pixelD = static_cast< PixType >( sum / kernel_width );
+          sum -= *pixelS1;
+          sum += *pixelS2;
+        }
+
+        // ending boundary case: the kernel is shrinking
+        for( i = kernel_width; i > half_width;
+             --i, pixelS1 += istepS, pixelD += istepD )
+        {
+          *pixelD = static_cast< PixType >( sum / i );
+          sum -= *pixelS1;
+        }
+      }
+    }
+  }
+
+  // Fast 1D (vertical) box filter smoothing
+  template < typename PixType >
+  void
+  box_average_vertical( const vil_image_view< PixType >& src,
+                        vil_image_view< PixType >& dst,
+                        unsigned kernel_height )
+  {
+    if( treat_as_interlaced )
+    {
+      // if interlaced, split the image into odd and even views
+      // transpose all input and ouput images so that the horizontal
+      // smoothing function produces vertical smoothing
+      vil_image_view< PixType > im_even_t =
+        vil_transpose( even_rows( src ) );
+      vil_image_view< PixType > im_odd_t =
+        vil_transpose( odd_rows( src ) );
+
+      vil_image_view< PixType > smooth_even_t =
+        vil_transpose( even_rows( dst ) );
+      vil_image_view< PixType > smooth_odd_t =
+        vil_transpose( odd_rows( dst ) );
+
+      // use a half size odd kernel since images are half height
+      int half_kernel_height = kernel_height / 2;
+      if( half_kernel_height % 2 == 0 )
+      {
+        ++half_kernel_height;
+      }
+      // Smooth transposed inputs with the horizontal smoothing
+      box_average_horizontal( im_even_t, smooth_even_t, half_kernel_height );
+      box_average_horizontal( im_odd_t, smooth_odd_t, half_kernel_height );
+    }
+    else
+    {
+      // if not interlaced, transpose inputs and outputs and apply horizontal
+      // smoothing.
+      vil_image_view< PixType > grey_img_t = vil_transpose( src );
+      vil_image_view< PixType > smooth_t = vil_transpose( dst );
+      box_average_horizontal( grey_img_t, smooth_t, kernel_height );
+    }
+  }
+
+  // Apply a high pass filter to one frame and return the output
+  template < typename PixType >
+  kwiver::vital::image_container_sptr
+  filter( vil_image_view< PixType >& input )
+  {
+    vil_image_view< PixType > output;
+    switch( mode )
+    {
+      case MODE_box:
+        output = box_high_pass_filter( input );
+        break;
+      case MODE_bidir:
+        output = bidirection_box_filter( input );
+        break;
+      default:
```

Comment:
Is an error or warning message needed here or is the `default` case effectively unreachable and only here to prevent compiler warnings?

Reply:
Yes, it's just for the compiler. Linus suggested using the `ENUM_CONVERTER` approach since it does quality error checking when there's a known set of values.